### PR TITLE
feat(api): add getText to ShapeUtil and text search example

### PR DESCRIPTION
**Pull request by MitjaBezensek**
Motivated by #3495 

An example of how to add custom search via`⌘+f` keyboard shortcuts. This also adds `getText` method to `ShapeUtil`, which allows the users to query the shapes for text.

Down the line we could choose to add something like `getShapesWithText` to the editor and maybe even add some UI to dotcom. For now, let's just expose the API.

https://github.com/user-attachments/assets/5631301c-a02a-4b66-8dbf-09571e67a46c

### Change type

- [x] `feature` (or improvement, bugfix, api, other - pick ONE)

### Test plan

1. Open the text search example.
2. Use ⌘+f to search for text within shapes.

- [ ] Unit tests (if present)
- [ ] End to end tests (if present)

### Release notes

- Adds `getText` to the `ShapeUtil` api so that we can allow searching for text in a nicely extensible way.
- Adds an example of how to add text search.

### CODE REVIEW

**Comment by mimecuvalo** [comment]
i love this so much

File: apps/examples/src/examples/text-search/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
+title: Text search
+component: ./TextSearchExample.tsx
+category: editor-api
+priority: 3
+keywords:
+  [
+    annotation,
```
Comment (mimecuvalo) [inline apps/examples/src/examples/text-search/README.md:8]: i think these keywords need updating maybe? the camera stuff doesn't seem related

File: apps/examples/src/examples/text-search/TextSearchExample.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { TLComponents, TLUiOverrides, Tldraw, atom, track } from 'tldraw'
+import 'tldraw/tldraw.css'
+import { TextSearchPanel } from './TextSearchPanel'
+import './text-search.css'
+
+export const showSearch = atom('showSearch', false)
+
+const components: TLComponents = {
+	HelperButtons: TextSearchPanel,
+	Minimap: null,
```
Comment (mimecuvalo) [inline apps/examples/src/examples/text-search/TextSearchExample.tsx:10]: does minimap need to be null?

File: packages/editor/src/lib/editor/shapes/ShapeUtil.ts [hunk 3]
```diff
@@ -373,6 +373,10 @@ export abstract class ShapeUtil<Shape extends TLUnknownShape = TLUnknownShape> {
 		return {}
 	}
 
+	getText(_shape: Shape): string | undefined {
```
Comment (mimecuvalo) [inline packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: since we're thinking of this new hook into the API, i think we should think of the general case where there is multiple text snippets, potentially, within a shape. so, i believe this should actually return:
`string[] | undefined` (and then the shape utils below should also return arrays).

_potentially_, that might mean it also returns more information? like id or position or something but i don't think we should get too complicated / over think this.

but i do think we should at least account for multiple textfields.

Comment (MitjaBezensek) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: I kinda like the lean approach. Until we see a use case I'd rather use the most simple api, which works for the current system. Stuff gets complicated enough already 😄

Comment (mimecuvalo) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: I feel ya on the lean approach. I just worry about the upgrade path here down the line. It feels like, yes, currently we only have one text field per shape. But this then doesn't become a usable API for other folks that have custom shapes (non-tldraw) that have multiple text fields. (there's feature requests that are asking for labels on both sides of an arrow for example). 

A different example would be how we have various checks in our geometry code to see if `isLabel` is true, because we don't want to bake in an assumption that there's only ever one label in the geometry. It keeps the optionality for multiple labels to exist open.

Comment (MitjaBezensek) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: I'm happy to change the api to return an array of strings instead of just a single one. As far as going further than that I don't feel confident in getting the API right without seeing more examples of how it could be used.

I don't think updating this in the future is that big of a problem. It's a fairly contained API and even if we do introduce breaking changes they should be quite limited in scope.

Comment (MitjaBezensek) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: Happy to let you take over if you understand the possible use cases better.

Comment (mimecuvalo) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: Yeah, some examples of how users could use it:
- tables: https://github.com/tldraw/tldraw/issues/1780
- arrows with multiple labels: https://github.com/tldraw/tldraw/issues/2141

And, again, we have to think about how we might have more primitive shapes (square, circle) on our board but our downstream users will have more complex shapes. So, another example could be, say, a slideshow creating tool. And there's a shape that let's you create title slides with Title, Subtitle, and a Date or something. Maybe they could compose that with separate text shapes individually each time but I could image that being annoying and having a combined Slide shape that encompasses multiple text labels like that.

But yeah, the table example above also shows a pretty good use case!

Comment (MitjaBezensek) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: It's still hypothetical. We don't have tables. Would you then expect the API to return an array of arrays for the tables?  Or just a flat array? Would it have some additional info associated with it (like positions)?
That's what I'm talking about when I say I don't understand the use case enough to be confident in coming up with an api.

Like I said I'm happy to change it from a single string to an array of strings if you feel that's better (I don't for now). But if you'd like more than that then you can either take this on or I can just close the PR if you feel it's not useful as is.

Comment (mimecuvalo) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: Hah, yeah, I know we don't have tables, I don't think we ever will in tldraw formally. But _other_ people will add tables to their usage of tldraw, and then they'll want to maybe use the search functionality.

> Would it have some additional info associated with it (like positions)?

Yes, I saw that above in my first comment: "potentially, that might mean it also returns more information? like id or position or something but i don't think we should get too complicated / over think this. but i do think we should at least account for multiple textfields."

Maybe we should talk offline at this point?  :)

Comment (mimecuvalo) [reply packages/editor/src/lib/editor/shapes/ShapeUtil.ts:376]: Talked offline and we'll go ahead land as is!
We'll naturally revisit all of this work anyway because of things like rich text, community feedback, and other feature requests.

**Comment by vickey-dev** [comment]
@MitjaBezensek thanks for introducing the search 🩷 , will it be part of [tldraw.com](https://www.tldraw.com/) because I couldn't find this on the tldraw website

**Comment by vickey-dev** [comment]
@MitjaBezensek  thanks for introducing the search 🩷 , will it be part of [tldraw.com](https://www.tldraw.com/) because I couldn't find this on the tldraw website

**Comment by MitjaBezensek** [comment]
> @MitjaBezensek thanks for introducing the search 🩷 , will it be part of [tldraw.com](https://www.tldraw.com/) because I couldn't find this on the tldraw website

Hey! We are considering adding it to tldraw.com, but don't have a short term plan of doing it. Would it be useful for you? would you expect it to work on a single page or across multiple pages? I guess you have complex documents with many shapes?

**Comment by vickey-dev** [comment]
Yes definitely useful for me @MitjaBezensek. For my case, I'm fine with the single page, I could also see the request across multiple pages as well https://github.com/tldraw/tldraw/issues/3495#issuecomment-2404665622 


>  I guess you have complex documents with many shapes?
Actually no, I mostly use the text & lines alone, I prefer to use it as a digital whiteboard where I jot down all the points related to the work I'm doing & with some docs links & flow charts.

**Comment by vickey-dev** [comment]
@MitjaBezensek  will this be part of any upcoming release to the tldraw.com
